code/modules/research/machinery/_production.dm

```diff
@@ -38,58 +44,133 @@
 /obj/machinery/rnd/production/Destroy()
 	materials = null
 	cached_designs = null
-	matching_designs = null
 	return ..()
 
+/// Updates the list of designs this fabricator can print.
 /obj/machinery/rnd/production/proc/update_designs()
 	cached_designs.Cut()
+
 	for(var/i in stored_research.researched_designs)
 		var/datum/design/d = SSresearch.techweb_design_by_id(i)
+
 		if((isnull(allowed_department_flags) || (d.departmental_flags & allowed_department_flags)) && (d.build_type & allowed_buildtypes))
 			cached_designs |= d
 
+	update_static_data(usr)
+
+	say("Successfully synchronized with R&D server.")
+
 /obj/machinery/rnd/production/RefreshParts()
 	. = ..()
 	calculate_efficiency()
 
-/obj/machinery/rnd/production/ui_interact(mob/user)
+/obj/machinery/rnd/production/ui_assets(mob/user)
+	return list(
+		get_asset_datum(/datum/asset/spritesheet/sheetmaterials),
+	)
+
+/obj/machinery/rnd/production/ui_interact(mob/user, datum/tgui/ui)
 	user.set_machine(src)
-	var/datum/browser/popup = new(user, "rndconsole", name, 460, 550)
-	popup.set_content(generate_ui())
-	popup.open()
 
+	ui = SStgui.try_update_ui(user, src, ui)
+
+	if(!ui)
+		ui = new(user, src, "Fabricator")
+		ui.open()
+
+/obj/machinery/rnd/production/ui_static_data(mob/user)
+	var/list/data = list()
+	var/list/designs = list()
+
+	for(var/datum/design/design in cached_designs)
+		var/cost = list()
+
+		for(var/datum/material/mat in design.materials)
+			var/coeff = efficient_with(design.build_path) ? efficiency_coeff : 1
+			cost[mat.name] = design.materials[mat] * coeff
+
+		designs[design.id] = list(
+			"name" = design.name,
+			"desc" = design.get_description(),
+			"cost" = cost,
+			"id" = design.id,
+			"categories" = design.category,
+			"maxstack" = design.maxstack
+		)
+
+	data["designs"] = designs
+	data["fab_name"] = name
+
+	return data
+
+/obj/machinery/rnd/production/ui_data(mob/user)
+	var/list/data = list()
+
+	data["materials"] = materials.mat_container?.ui_data()
+	data["on_hold"] = materials.on_hold()
+	data["busy"] = busy
+
+	return data
+
+/obj/machinery/rnd/production/ui_act(action, list/params)
+	. = ..()
+
+	if(.)
+		return
+
+	. = TRUE
+
+	switch (action)
+		if("remove_mat")
+			var/datum/material/M = locate(params["ref"])
```

Comment:
1. Full name
2. Validate that this is actually a valid material inside this lathe. An exploiter can send any ref they want in here.

Reply:
Interestingly, this is exactly how fabs currently handle this--I copied this 1:1. I'll see what I can do.